### ISSUE TITLE
Getting DEF file during build process

### ISSUE THREAD

**Issue by 7heW4yne** [aubio/aubio#126]
I mentioned in #114 that it seems to be possible in the newer waflib versions to get a valid def file using the syms extra.

Using my [aubio-Vagrant-Script](https://github.com/7heW4yne/aubio-Vagrant-Script) I have a ubuntu linux with mingw and I'm trying to port the [changes for the first syms extension](https://gist.github.com/7heW4yne/c707f78f63fa301f173fc1d5f19d3199) into the current aubio sources, in order to get a valid DEF file with all builds (not only mingw).

But I get into trouble when I try to get waf.
@piem updated the get-waf.sh to version 1.9.12. After that I call `make expandwaf` and get the error:

> Makefile:51: recipe for target 'expandwaf' failed
make: *** [expandwaf] Error 2`

I looked in the Makefile and tried to do the steps manually.
And I found out that `python waf --help` gives following error:

> Traceback (most recent call last):
  File "/aubio/.waf-1.9.12-46ae1733752575b7ea2378b1eca4e1ae/waflib/Scripting.py", line 122, in waf_entry_point
    run_commands()
  File "/aubio/.waf-1.9.12-46ae1733752575b7ea2378b1eca4e1ae/waflib/Scripting.py", line 179, in run_commands
    parse_options()
  File "/aubio/.waf-1.9.12-46ae1733752575b7ea2378b1eca4e1ae/waflib/Scripting.py", line 152, in parse_options
    Context.create_context('options').execute()
  File "/aubio/.waf-1.9.12-46ae1733752575b7ea2378b1eca4e1ae/waflib/Options.py", line 149, in execute
    super(OptionsContext,self).execute()
  File "/aubio/.waf-1.9.12-46ae1733752575b7ea2378b1eca4e1ae/waflib/Context.py", line 88, in execute
    self.recurse([os.path.dirname(g_module.root_path)])
  File "/aubio/.waf-1.9.12-46ae1733752575b7ea2378b1eca4e1ae/waflib/Context.py", line 129, in recurse
    user_function(self)
  File "/aubio/wscript", line 100, in options
    ctx.load('compiler_c syms')
  File "/aubio/.waf-1.9.12-46ae1733752575b7ea2378b1eca4e1ae/waflib/Context.py", line 82, in load
    module=load_tool(t,path,with_sys_path=with_sys_path)
  File "/aubio/.waf-1.9.12-46ae1733752575b7ea2378b1eca4e1ae/waflib/Context.py", line 389, in load_tool
    __import__(tool)
ImportError: No module named syms

I checked the waf system and as far as I can see, the downloaded waf file (get_waf.sh) contains the zipped waflib data at its end (block starting with `==>` and ending with `<==`). Calling `python waf --help` triggers the expansion of these data and creates the folder `.waf-1.9.12-46ae1733752575b7ea2378b1eca4e1ae/waflib/`. This folder should contain the complete waflib for version 1.9.12. **BUT** instead its `extras` folder is almost empty. And I think that's the reason for the missing syms module.

I also tried with waf 1.9.6 and 1.9.13, but the `extras` folder is always incomplete.

Any ideas how to change the `get_waf.sh` or Makefile in a way, that the waflib folder is updated [including all files as in the waflib repository](https://github.com/waf-project/waf/tree/waf-1.9.12/waflib/extras)?

**Comment by piem** [comment]
hi @7heW4yne and thanks for your work and input on #114.

the problem is the waf script downloaded by aubio's makefile does not containt the syms module.

you should be able to build one following the instsructions @ [waf.io](https://waf.io/book/#_customization_and_redistribution):

    python waf-light --tools=syms

**Comment by 7heW4yne** [comment]
Hi @piem 

Ok. This sounds good. I will try it asap and post back.

I also started my own fork of aubio for that syms stuff. And I opened an issue at the wag project, because there seems to be a problem in the syms extra using our regular expression for the exported symbols.

I hope, I will get some results next week. Perhaps this brings us a good solution for the def files and maybe also an import lib for windows.

**Comment by 7heW4yne** [comment]
Last night, I found a solution for the exception in the `syms` extra of waf. The problem was our regular expression and a not documented restriction of the `syms.py` (also see [issue #2007 @ waf](https://github.com/waf-project/waf/issues/2007)).

I also did some tests with the `python waf-light --tools=syms` command. It seems that the instructions from the waf book @ waf.io should be handled with care (same to the waf support), because the command in the waf book is wrong. If I call it, I get "error: no such option: --tools".
Instead it has to be:
`python waf-light configure build --tools=syms`

Anyway, I try to change the `get_waf.sh` in my fork now. BUT first, I want to ask you (@piem) for your opinion, because there are 2 options to get waf+syms.

The official way (Option 1) would be:
1. Download and unpack the waf-1.9.13.**tar.bz2** (instead of the pre-packed waf file without extras)
2. Call `python waf-light configure build --tools=syms`
3. Copy the created waf file to aubio root folder
4. Continue [here](https://github.com/aubio/aubio/blob/master/Makefile#L51)

The short way (Option 2) would be:
1. Download and unpack the waf-1.9.13.tar.bz2
2. Copy the whole waflib folder (with all extras and tools) to the aubio root folder.
3. move `waf-light` to `/<aubio-root>/waf`
4. Continue [here](https://github.com/aubio/aubio/blob/master/Makefile#L51)

I think, in order to be compatible with future versions of waf, the official way (Option 1) should be preferred. What do you think @piem?

**Comment by 7heW4yne** [comment]
@piem seems to be absent at the moment. That's why I worked on according to my point of view.

Up to now, I have in [my aubio fork](https://github.com/7heW4yne/aubio):
1. Updated to waf 1.9.13
2. Changed get_waf.sh (official way to have future compatibility)
3. Create the def file using the syms extra
4. Updated build_mingw script
5. Added def file to `waf install` command (copy the created def file to lib folder

Now I only have 1 problem remaining.
For win64 the created def file is empty. And this seems to be related to the combination of the win64 mingw gcc + the syms extra of waf. To find a solution I opened [issue 2008 @ waf
](https://github.com/waf-project/waf/issues/2008).

If this problem can be fixed, I intend to start a pull request, but I don't know how to "pass all continuous integration tests". I hope @piem can give me some advice on that when he is back :)

**Comment by 7heW4yne** [comment]
The issue @ waf is closed. I think, I found a solution. But it will not get integrated in waf for the following reasons:
1. If it would, I would have to change the `get_waf.sh` again, to get the current sources from the github project (we could not use the prepared archives from https://waf.io anymore). But this would increase the risk of errors, because the current github sources may contain experimental code ... especially the "extras" as the owner of waf recently told me (just read issue 2008 @ waf).
2. I don't want to spend my time anymore to help the waf project, because of its owner.

I gave waf a try, but it's easier to use the solution we already have in `build_mingw`. That's also why I stop working on my aubio fork.

Thanks @piem for your time and support on that.
